### PR TITLE
Bump stylish-haskell and use new import grouping feature

### PR DESCRIPTION
**Pull request by michaelpj**
- Bump stylish-haskell to 0.14.4
- Add grouping to stylish-haskell config
- Apply changes

EDIT: hmm, my PR body got lost. This bumps our `stylish-haskell` version so we can group imports in the way we've often discussed doing, but automatically. The rule mechanism is fairly flexible, so we can do various things quite easily. I think what I've put here is decent, namely:
- Group everything with "Plutus" in the first component
- Within that, group on the first component
- Everything else goes together

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

**Review by thealmarty — APPROVED** [review]
So tidy! I like it! Maybe worth a change log?

**Comment by michaelpj** [comment]
> Maybe worth a change log?

It's not user-facing?

File: plutus-core/plutus-core/src/PlutusCore/Subst.hs [hunk 1]
```diff
@@ -24,10 +24,10 @@ module PlutusCore.Subst
     , tvTy
     ) where
 
-import PlutusPrelude
-
 import PlutusCore.Core
 
+import PlutusPrelude
+
 import Control.Lens
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Subst.hs:31]: Meh.

File: plutus-core/plutus-core/test/Names/Spec.hs [hunk 2]
```diff
@@ -7,17 +7,15 @@
...
 import PlutusCore
 import PlutusCore.DeBruijn
-import PlutusCore.Mark
-import PlutusCore.Pretty
-import PlutusCore.Rename.Internal
-
 import PlutusCore.Generators.Hedgehog
 import PlutusCore.Generators.Hedgehog.AST as AST
 import PlutusCore.Generators.Hedgehog.Interesting
+import PlutusCore.Mark
```
Comment (effectfully) [inline plutus-core/plutus-core/test/Names/Spec.hs:15]: No.

File: plutus-core/plutus-ir/src/PlutusIR/Parser.hs [hunk 3]
```diff
@@ -14,18 +14,20 @@ module PlutusIR.Parser
...
 import PlutusIR.MkPir qualified as PIR
+
 import PlutusPrelude
-import Prelude hiding (fail)
 
 import Control.Monad.Combinators.NonEmpty qualified as NE
 import Control.Monad.Except (MonadError)
 import Data.Text (Text)
-import PlutusCore (MonadQuote)
-import PlutusCore.Error (AsParserErrorBundle)
```
Comment (effectfully) [inline plutus-core/plutus-ir/src/PlutusIR/Parser.hs:28]: It clearly does some good stuff too.

File: plutus-core/plutus-ir/src/PlutusIR/Parser.hs [hunk 4]
```diff
@@ -14,18 +14,20 @@ module PlutusIR.Parser
...
+import Prelude hiding (fail)
 import Text.Megaparsec hiding (ParseError, State, many, parse, some)
```
Comment (effectfully) [inline plutus-core/plutus-ir/src/PlutusIR/Parser.hs:31]: But then.

File: plutus-core/plutus-ir/src/PlutusIR/Purity.hs [hunk 5]
```diff
@@ -5,10 +5,11 @@
 
 module PlutusIR.Purity (isPure, firstEffectfulTerm) where
 
+import PlutusCore.Builtin
+
 import PlutusIR
```
Comment (effectfully) [inline plutus-core/plutus-ir/src/PlutusIR/Purity.hs:10]: It also breaks our policy of most "distant" modules going last.

File: plutus-core/plutus-ir/test/GeneratorSpec.hs [hunk 6]
```diff
@@ -5,7 +5,6 @@ import GeneratorSpec.Builtin
 import GeneratorSpec.Substitution
 import GeneratorSpec.Terms
 import GeneratorSpec.Types
-
 import Test.QuickCheck
```
Comment (effectfully) [inline plutus-core/plutus-ir/test/GeneratorSpec.hs:8]: Why?

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Result.hs [hunk 7]
```diff
@@ -16,10 +16,10 @@ module PlutusCore.Evaluation.Result
     , isEvaluationFailure
     ) where
 
-import PlutusPrelude
-
 import PlutusCore.Pretty
 
+import PlutusPrelude
+
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Result.hs:22]: This is pretty much the only thing it does in modules I've written, i.e. only makes things worse. Hard for me to vote in favor of adopting it.

File: plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs [hunk 8]
```diff
@@ -27,13 +27,12 @@ import PlutusCore.Normalize.Internal (MonadNormalizeType)
...
 import PlutusCore.Rename
+
 import PlutusPrelude
 
 import Control.Lens
 import Control.Monad.Error.Lens
 import Control.Monad.Except
--- Using @transformers@ rather than @mtl@, because the former doesn't impose the 'Monad' constraint
--- on 'local'.
 import Control.Monad.Trans.Reader
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs:36]: And deletes comments??? OK, that's enough.

**Review by effectfully — CHANGES_REQUESTED** [review]
image

**Comment by effectfully** [comment]
Maybe it's configurable? If it wasn't removing newlines, removing comments, reordering import groups, randomly assigning modules without prefixes to new groups...

**Comment by michaelpj** [comment]
> Maybe it's configurable?

It's pretty configurable. You just provide it some regexes to use to group things, and it groups them. You can see the config I put in the `.hlint.yaml` file. If you can articulate what you want it to do I can see if I can make it do it! What I've put currently is my understanding of what we "usually" do.

File: plutus-core/plutus-ir/src/PlutusIR/Purity.hs (hunk 5 above)
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Purity.hs:10]: If we have such a policy I don't know it :sweat_smile: that's partly why I want to try and codfiy it!

**Comment by michaelpj** [comment]
Looking at your comments, I think you want a rule like: if there are any `X.Y.Z` modules, then make a subgroup for the `X.Y` prefix, otherwise group it under `X`?

So you want

```
PlutusCore.X
PlutusCore.Y

PlutusCore.Z.P
PlutusCore.Z.Q
```
?

**Comment by effectfully** [comment]
> Looking at your comments, I think you want a rule like: if there are any X.Y.Z modules, then make a subgroup for the X.Y prefix, otherwise group it under X?

Ideally yes, but not merging existing groups would also be (while allowing individual imports to switch groups).

**Comment by michaelpj** [comment]
I don't think we can configure it to respect existing groups, and indeed part of the point is I'd like to not have to think about import grouping at all.

**Comment by michaelpj** [comment]
Sadly many formatters have comment problems :( looks like this one is old: https://github.com/haskell/stylish-haskell/issues/39

**Pull request closed without merging**